### PR TITLE
The client now does not receive bitrate information

### PR DESCRIPTION
**Pull request by saurpath**

### CODE REVIEW

File: starter_proxy/src/proxy.c [hunk 1]
```diff
@@ -338,25 +338,45 @@ int process_client_read(client **clients, size_t i, int data_available,
...
+                    dbg("[proxy.process_client_read] New request creation "
+                        "failed. Exiting.\n");
+                    exit(EXIT_FAILURE);
+                }
+
+                int bytes_queued =
+                    queue_message_send(clients, i, new_query, msg_length);
+                // Mark proxy connection ready to be written to
+                FD_SET(clients[i]->fd, write_set);
+                clients[sibling_idx]->manifest_request = NULL;  // reset
```
Comment (jkosh44) [inline starter_proxy/src/proxy.c:375]: We should free this before setting it to NULL to avoid memory leaks

File: starter_proxy/src/proxy.c [hunk 2]
```diff
@@ -338,25 +338,45 @@ int process_client_read(client **clients, size_t i, int data_available,
...
                 char *body = get_response_body(msg_rcvd, msg_length);
                 if (parse_manifest(clients, sibling_idx, body)) {
                     // Parsing failed
                     dbg("[proxy.process_client_read] Parsing failed. "
-                        "Existing.\n");
+                        "Exiting.\n");
                     exit(EXIT_FAILURE);
                 }
                 print_client_pool(clients);
+                // TODO: Make another query to server for no_list.
```
Comment (jkosh44) [inline starter_proxy/src/proxy.c:358]: We can probably remove this now.

File: starter_proxy/src/proxy.c [hunk 3]
```diff
@@ -338,25 +338,45 @@ int process_client_read(client **clients, size_t i, int data_available,
...
+
+                // Create modified query
+                char *new_query = get_nolist_req(
+                    clients[sibling_idx]->manifest_request, &msg_length);
```
Comment (jkosh44) [inline starter_proxy/src/proxy.c:362]: It doesn't seem right that we pass in `msg_length` here. Isn't `message_length` the length of the manifest response file and not the length of the manifest HTTP request?

File: starter_proxy/src/utils.c [hunk 4]
```diff
@@ -165,6 +182,44 @@ void update_manifest_record(client **clients, int i, char *media_tag) {
...
+ */
+char *get_nolist_req(char *request, size_t *req_len) {
+    char *uri, *filename, *new_uri, *body;
+
+    if ((uri = get_request_uri(request)) != NULL) {
+        filename = get_filename(uri);
+        if (filename == NULL) return NULL;
+
+        // Increase msg-length for _nolist
+        *req_len = *req_len + 7;
```
Comment (jkosh44) [inline starter_proxy/src/utils.c:202]: Re my comment above: It seems like we're adding 7 to the length of the manifest response file and not the manifest HTTP request.

File: starter_proxy/src/utils.c [hunk 5]
```diff
@@ -165,6 +182,44 @@ void update_manifest_record(client **clients, int i, char *media_tag) {
...
+        new_uri = calloc(1, MAX_FILE_LEN);
+        snprintf(new_uri, *req_len, "%s_nolist.f4m", filename);
+
+        request = realloc(request, *req_len);
+        // Assuming all requests are HTTP1. requests.
+        body = strstr(request, "HTTP/1");  // Get old body
+        dbg("[utils.get_nolist] Request body:\n%s", body);
+
+        // Create new request
+        snprintf(request, *req_len, "GET %s %s", new_uri, body);
```
Comment (jkosh44) [inline starter_proxy/src/utils.c:213]: I think the error might be somewhere in here. I'm not sure how safe it is to call `snprintf` with overlapping strings. i.e. `body` is part of `request` and you're overwriting part of `body` with `request`. 

So if `new_uri` overwrites part of the `body` array, then when you go to copy `body`, you've lost some data.
For example let's say `request` is length 100, `body` is starts at index 10 in `request` and is length 90, and `new_uri` is length 15. Then first `new_uri` is copied into `request` which overwrites the first 5 characters of `body`. Then when you copy `body` into `request` you've lost the first 5 characters and will double copy the last 5 characters of `new_uri`.

I think it'd be much safer to just create a new string.

File: starter_proxy/src/utils.c [hunk 6]
```diff
@@ -165,6 +182,44 @@ void update_manifest_record(client **clients, int i, char *media_tag) {
...
+ * purpose: Creates a request for _nolist manifest file from original manifest
+ * file.
+ *
+ * returns: Updated request buffer for nolist file. NULL for invalid requests.
+ */
+char *get_nolist_req(char *request, size_t *req_len) {
+    char *uri, *filename, *new_uri, *body;
+
+    if ((uri = get_request_uri(request)) != NULL) {
+        filename = get_filename(uri);
```
Comment (jkosh44) [inline starter_proxy/src/utils.c:198]: I'm pretty sure URI is NOT a null terminated string here, but `get_filename` will call `get_file_extension` which calls `strrchr` which expects a null terminated string.

Comment (jkosh44) [reply starter_proxy/src/utils.c:198]: This same issue actually exists in `is_manifest` too.

File: starter_proxy/src/proxy.c (hunk 1 above)
Comment (saurpath) [reply starter_proxy/src/proxy.c:375]: It just a pointer to request we got. I never allocated memory to it. The free for final buffer should take of it.

File: starter_proxy/src/proxy.c (hunk 3 above)
Comment (saurpath) [reply starter_proxy/src/proxy.c:362]: Didn't notice. Good catch.

File: starter_proxy/src/proxy.c [hunk 7]
```diff
@@ -368,17 +374,22 @@ int process_client_read(client **clients, size_t i, int data_available,
...
-                    queue_message_send(clients, i, new_query, msg_length);
+                bytes_queued = queue_message_send(
+                    clients, i, new_query,
+                    clients[sibling_idx]->manifest_request_len);
                 // Mark proxy connection ready to be written to
                 FD_SET(clients[i]->fd, write_set);
-                clients[sibling_idx]->manifest_request = NULL;  // reset
+
+                // reset
+                clients[sibling_idx]->manifest_request = NULL;
```
Comment (jkosh44) [inline starter_proxy/src/proxy.c:384]: On line 346 - 347 we call 
```C
clients[i]->manifest_request = calloc(1, msg_length);
memcpy(clients[i]->manifest_request, msg_rcvd, msg_length);
```

but I don't see where we ever free that memory allocated by calloc.

**Review by jkosh44 — APPROVED** [review]

File: starter_proxy/src/proxy.c (hunk 7 above)
Comment (saurpath) [reply starter_proxy/src/proxy.c:384]: This gets free when we call queue_message_send.

The flow would be calloc here -> realloc in get_nolist_req -> freed in queue_message_send

**Pull request merged**